### ISSUE TITLE
Query Functionality with `onCompleted` is broken in v3.0.0 release (CodeSandbox repro)

### ISSUE THREAD

**Issue by 3nvi** [apollographql/apollo-client#6636]
Some core lazy querying functionality through `useLazyQuery` seems a bit off.

Specifically, when a lazy query (through a `useLazyQuery` hook) returns an  error, then Apollo seems to start **endlessly** querying the server. The same thing happens if you give it a `fetchPolicy`  of `network-only` (or anything that doesn't search the cache first).

Related: https://github.com/apollographql/react-apollo/issues/4044

At the same time, if the lazy query function (returned by `useLazyQuery`) gets called more than 1 times, then no callbacks are fired. The `onCompleted` doesn't fire again. The `onCompleted` fires on the 1st successful run of the "query" function and then consecutive invocations of the lazy query function, don't trigger `onCompleted` callback (they should, since you can't have conditional business logic based on whether the data  is cached or not!)

Related: https://github.com/apollographql/react-apollo/issues/4000

Funnily enough, using `client.query()` works fine. All this used to work perfectly during the beta (I think I've verified that something occurred after `@apollo/client@v3.0.0-beta.43`),  but now it doesn't.

UPDATE: I've  added a codesandbox that  reproduces it
 
**Intended outcome:**
- Lazy query errors shouldn't trigger endless API querying
- A `network-only` fetch policy  shouldn't trigger endless API querying on a lazy query
- `onCompleted` should fire every time a lazy query gets successfully executed

**Actual outcome:**
- Lazy query errors & `network-only` fetch policy start an endless loop of requests
- `onCompeted` gets fired at most once regardless of the number of invocations

**How to reproduce the issue:**

- Take a  lazy query and change the `fetchPolicy` to `network-only` **or** force an API error on a lazy query
- Click a button that triggers a lazy query multiple time, while observing the number of invocations of the `onCompleted` callback.

**Versions**
  System:
    OS: macOS 10.15.2
  Binaries:
    Node: 12.18.0 - /usr/local/bin/node
    Yarn: 1.19.0 - /usr/local/bin/yarn
    npm: 6.14.4 - /usr/local/bin/npm
  Browsers:
    Chrome: 83.0.4103.116
    Firefox: 72.0.2
    Safari: 13.0.4
  npmPackages:
    @apollo/client: ^3.0.2 => 3.0.2
    apollo-link-error: ^1.1.12 => 1.1.13

**Comment by 3nvi** [comment]
@benjamn  Take a look here and observe the console

https://codesandbox.io/s/apollo-reference-equality-1exes?file=/src/App.js

While `fetchPolicy` is `network-only` React re-renders endlessly. If you make it `cache-first` things are ok again. Same thing would happen if the query errored.

Also observe how the `ON COMPLETED` gets only logged once

**Comment by 3nvi** [comment]
I noticed that even if I change the `variables` after one query execution, the query automatically  executes (although it's lazy, so it shouldn't).

In general, this seems like a  bug that may be unrelated to lazy-loading, but I feel it's something on the "core" functionality of querying since it  manifests in many different ways

**Comment by 3nvi** [comment]
Related to https://github.com/apollographql/react-apollo/issues/4044

**Comment by 3nvi** [comment]
Also related to https://github.com/apollographql/react-apollo/issues/3729

**Comment by kainanaina** [comment]
I'm using `3.0.0-beta.43` since February and everything worked there just fine. Tried an upgrade to stable v3 and encountered this issue.

**Comment by orrybaram** [comment]
I've noticed that this is also happening with a regular `useQuery`  hook when an `onCompleted` function is specified (even if its a noop) and the `fetchPolicy` is set to anything except `cache-first` or `cache-only`

**Comment by 3nvi** [comment]
> I've noticed that this is also happening with a regular `useQuery` hook when an `onCompleted` function is specified (even if its a noop) and the `fetchPolicy` is set to anything except `cache-first` or `cache-only`

Yeah I also mentioned that I feel that  this is not necessarily tied only to lazy queries.  Will update description

**Comment by Marcelo-Figueiredo** [comment]
Same problem. Endless request loop when using onCompleted in one of my applications after upgrading. Might go back to apollo-boost for now.

**Comment by hwillson** [comment]
This issue was fixed in https://github.com/apollographql/apollo-client/pull/6588, and will be coming in `@apollo/client@3.1.0` very soon.

**Comment by benjamn** [comment]
Heads up: we're not quite ready to publish `@apollo/client@3.1.0` yet, but I've published an `@apollo/client@3.1.0-pre.0` release that includes #6588, if you want to try that in the meantime. 🙏

**Comment by 3nvi** [comment]
> Heads up: we're not quite ready to publish `@apollo/client@3.1.0` yet, but I've published an `@apollo/client@3.1.0-pre.0` release that includes #6588, if you want to try that in the meantime. 🙏
> 
> Update: thanks to @3nvi's reproduction, I can confirm the fix: https://codesandbox.io/s/apollo-reference-equality-t0ttx

Perfect.  Thanks a lot @benjamn !

**Comment by emjaksa** [comment]
@benjamn I just tried the latest version ``@apollo/client@3.0.2`` and `` @apollo/client@3.1.0-pre.0``. When I try execute the search below the query gets stuck in an infinite loop, making multiple network requests and never calling onComplete. If I comment out either the fetchPolicy or onCompleted it stops but the combination of both fetch ``fetchPolicy: 'no-cache'`` and ``onCompleted: data => {}`` cause the query to go into an infinite loop.

I tried replicating it in your sandbox but I can only replicate it with ``@apollo/client@3.0.2``.

```
const [search] = useLazyQuery(SearchQuery, {
    fetchPolicy: 'no-cache',
    onCompleted: data => {},
  })

const handleClick = () => { 
  search({
    variables: {
       query: 'search',
    }
  }) 
}

```

**Comment by Tautorn** [comment]
> @apollo/client@3.1.0-pre.0

Thank you so much @benjamn ! It's working for me, great!

**Comment by cyberwombat** [comment]
I just tried @apollo/client@3.1.0-pre.0 and it's still a no go. This issue might be slightly different than the 2 year old [one](https://github.com/apollographql/react-apollo/issues/2177) (auto archived without fix and reopened by me https://github.com/apollographql/react-apollo/issues/3968 then repo archived). But the gist for me is that `onCompleted` doesn't fire again rendering the concept of a cache useless.  I am a bit confused how anyone is able to use a cache as we never have been able to. @Tautorn you say its working for you - is the cache enabled? how are you adding business logic when refetching from cache?

**Comment by emjaksa** [comment]
@apollo/client@3.1.0-pre.0 worked for me. I had to delete the ``package-lock.json`` and run ``npm install`` again. Very odd

**Comment by Tautorn** [comment]
> @benjamn I just tried the latest version `@apollo/client@3.0.2` and ` @apollo/client@3.1.0-pre.0`. When I try execute the search below the query gets stuck in an infinite loop, making multiple network requests and never calling onComplete. If I comment out either the fetchPolicy or onCompleted it stops but the combination of both fetch `fetchPolicy: 'no-cache'` and `onCompleted: data => {}` cause the query to go into an infinite loop.
> 
> I tried replicating it in your sandbox but I can only replicate it with `@apollo/client@3.0.2`.
> 
> ```
> const [search] = useLazyQuery(SearchQuery, {
>     fetchPolicy: 'no-cache',
>     onCompleted: data => {},
>   })
> 
> const handleClick = () => { 
>   search({
>     variables: {
>        query: 'search',
>     }
>   }) 
> }
> ```

Hi there.

The cache is enabled, but I don't use `fetchPolicy` to get any result.

I'm using `onCompleted` after query. Before I updated to `@apollo/client@3.1.0-pre` was necessary wrapper `useQuery` and `useLazyQuery` with `useCallback`.

But now working normally (after update version).

**Comment by zeeshanaligold** [comment]
```
const { loading, error } = useQuery(GET_USER, {
    onCompleted({ user }) {
      setAuth({ user, isAuthenticated: true })
    },
})
```
Still getting stuck in infinite loop of requests to api. it's happening only when i call setAuth to save state.

**Comment by zeeshanaligold** [comment]
```
const { loading, error } = useQuery(GET_USER, {
    onCompleted({ user }) {
      setAuth({ user, isAuthenticated: true })
    },
})
```
Still getting stuck in infinite loop of requests to api. it's happening only when i call setAuth to save state.

**Comment by zeeshanaligold** [comment]
```
const { loading, error } = useQuery(GET_USER, {
    onCompleted({ user }) {
      setAuth({ user, isAuthenticated: true })
    },
})
```
Still getting stuck with infinite loop of requests to api. it's happening only when i call setAuth to save state with context api.

**Comment by orrybaram** [comment]
@zeeshanaligold try adding a flag to check for the user in the `onCompleted` method. You may be accidentally retriggering the infinite loop yourself by setting state on each render (which cause it to re-render again)

```
onCompleted({ user }) {
  if (!authedUser) {
    setAuth({ user, isAuthenticated: true })
   }    
}
```

**Comment by Tautorn** [comment]
> ```
> const { loading, error } = useQuery(GET_USER, {
>     onCompleted({ user }) {
>       setAuth({ user, isAuthenticated: true })
>     },
> })
> ```
> 
> Still getting stuck with infinite loop of requests to api. it's happening only when i call setAuth to save state with context api.

@zeeshanaligold try with `useCallback`:

```jsx
import { useCallback } from 'react'
...

const handleOnCompleted = useCallback(({ user }) => {
    setAuth({ user, isAuthenticated: true })
}, [])

const { loading, error } = useQuery(GET_USER, {
    onCompleted: handleOnCompleted
})
```

**Comment by zeeshanaligold** [comment]
> @zeeshanaligold try adding a flag to check for the user in the `onCompleted` method. You may be accidentally retriggering the infinite loop yourself by setting state on each render (which cause it to re-render again)
> 
> ```
> onCompleted({ user }) {
>   if (!authedUser) {
>     setAuth({ user, isAuthenticated: true })
>    }    
> }
> ```

@orrybaram it's fix the infinite loop issue but still sending 2 requests.

**Comment by zeeshanaligold** [comment]
> ```js-jsx
> const handleOnCompleted = useCallback(({ user }) => {
>     setAuth({ user, isAuthenticated: true })
> }, [])
> ```

@Tautorn  i did try it but still same issue.

**Comment by zeeshanaligold** [comment]
> ```js-jsx
> import { useCallback } from 'react'
> ...
> 
> const handleOnCompleted = useCallback(({ user }) => {
>     setAuth({ user, isAuthenticated: true })
> }, [])
> 
> const { loading, error } = useQuery(GET_USER, {
>     onCompleted: handleOnCompleted
> })
> ```

@Tautorn i did try it but still same issue.

**Comment by Tautorn** [comment]
@zeeshanaligold Could you reproduction this issue in some tool like codesandbox? If possible I hope.

**Comment by FezVrasta** [comment]
Does 3.1.1 include this fix?

**Comment by GeorgeMarcus** [comment]
Not fixed on 3.1.1. It still only works by wrapping `onCompleted` and `onError` in `useCallback`.

**Comment by emjaksa** [comment]
3.1.1 is working for me

**Comment by 3nvi** [comment]
> Not fixed on 3.1.1. It still only works by wrapping `onCompleted` and `onError` in `useCallback`.

Ι haven't tested it yet, but make sure the you reset this  particular `node_module` and that `package-lock.json` correctly references the installed version

**Comment by 3nvi** [comment]
@benjamn and @hwillson Thanks for tackling this. 

## Current State of things

As of the moment of writing, in `3.1.1`:

- [x] An errorred query no longer re-renders endlessly  🎉 
- [x] A `cache-first` query with `onCompleted` and/or `onError` no longer re-renders endlessly 🎉 
- [x] A `network-only`  query  no longer pings the server endlessly 🎉 

## Pending bugs

- [ ] The `onCompleted` callback fires whenever a query is resolved (either from network or cache). https://github.com/apollographql/react-apollo/issues/4000

> *Currently only `network-only` requests have their `onCompleted` fired after every query execution. For `cache-first` ones, the `onCompleted` fires only once. Unless this is documented somewhere, I would flag this as a bug since our `onCompleted` logic shouldn't have to account for the origin of this data. It should be  agnostic of that*

- [ ] Changing a query variable on a lazy query automatically fires it. https://github.com/apollographql/react-apollo/issues/3729

> *This should definitely happen for  `useQuery` hooks, but **not** for `useLazyQuery` hooks (since they wouldn't  be lazy anymore). Again if this is the expected functionality it should b documented*


## Reproduction of pending bugs

You can observe both these bugs in the  following codesandbox --> https://codesandbox.io/s/apollo-reference-equality-op12z?file=/src/App.js

-  **To reproduce the `onCompleted` error**: Change `fetchPolicy` to `cache-first`, click "Run Query" and see that the `onCompleted` doesn't fire

- **To reproduce the `variables` with `useLazyQuery` error**: Click "Run Query" once and then start clicking "Change foo". Clicking "change foo" should **not** cause a query to run

### Additional Context

Out of the two bugs listed above (in the order that they are mentioned):

- The first  is within the scope of this ticket and a requirement in order to close it 
- The second isn't necessarily related to this ticket (although mentioned)and I understand that it may require a more complex solution with observables. Thus, I'm willing to port (2) to another issue in order to help eventually close the existing one

Thanks again!

**Comment by uncvrd** [comment]
I don't think I've seen this mentioned yet as it probably comes from the same underlying behavior...but a `useLazyQuery` with a `onCompleted` does not get called again if a user specifies a `refetchQueries` array in a `useMutation` method call. 

Such as 

```
const [updateAlbumReadStatus] = useMutation({
        // this triggers a network refetch but does not call the onComplete() of my UnreadAlbums useLazyQuery
        refetchQueries: ["UnreadAlbums"]
    });
```

**Comment by lenarmazitov** [comment]
Attention!

Problem still exists in 3.1.1
for fetchPolicy: "cache-and-network"

(with fetchPolicy: "network-only" all works fine)

**Comment by warent** [comment]
I'm on 3.1.1 and onCompleted never calls if a useLazyQuery has its result cached.

**Comment by benjamn** [comment]
Those of you having problems with `cache-and-network` or `network-only` making unwanted network requests (especially if `cache-first` seems to fix things), please read this comment about the new `options.nextFetchPolicy` option introduced in v3.1.0: https://github.com/apollographql/apollo-client/issues/6760#issuecomment-668188727

**Comment by finchen** [comment]
I had the same issue with "@apollo/client": "^3.1.3" (loading always `false` and no callback).

I found out that the `variables` must use react state.
Following code keeps `loading` always `true` and never goes to `onCompleted` callback:
```js
const { loading, error, data } = useQuery(LOGS, {
  variables: {
      dateFrom: new Date().toISOString()
    }
}
```
Following code works but still have to add `notifyOnNetworkStatusChange: true` for the `onCompleted` callback:
```js
const [dateFrom, setDateFrom] = useState(new Date().toISOString());
const { loading, error, data } = useQuery(LOGS, {
  notifyOnNetworkStatusChange: true,
  variables: {
      dateFrom
    }
}
```

**Comment by gtb104** [comment]
I solved my specific issue. I was creating the client within the component, so it was recreating the client each time the results were returned. I changed it to:
```javascript
  const client = useRef(getClient());

  const { data, error, loading } = useQuery(gql(queries.listDevices), {
    client: client.current,
    variables: {
      tenantId: tenantId()
    }
  });
```
The `getClient` function does the `new ApolloClient` stuff from above.

**Comment by benjamn** [comment]
@gtb104 There's a context hook for that! See [`useApolloClient`](https://www.apollographql.com/docs/react/api/react/hooks/#useapolloclient), which works if you're using [`ApolloProvider`](https://www.apollographql.com/docs/react/api/react/hooks/#apolloprovider) somewhere further up your component tree.

**Comment by nodabladam** [comment]
I just discovered that after upgrading to apollo 3 that my `useLazyQuery` are not always firing the onCompleted.  Current workaround for me seems to be changing usage of useLazyQuery to:
```
client.query({ query: <QUERY> }).then(res =>{ ... } );
```

**Comment by adrienharnay** [comment]
> Those of you having problems with `cache-and-network` or `network-only` making unwanted network requests (especially if `cache-first` seems to fix things), please read this comment about the new `options.nextFetchPolicy` option introduced in v3.1.0: [#6760 (comment)](https://github.com/apollographql/apollo-client/issues/6760#issuecomment-668188727)

For this bug: `Changing a query variable on a lazy query should not automatically fire it.`, I don't think `options.nextFetchPolicy` can be of any help, and I believe there's no workaround.

**Comment by benjamn** [comment]
@3nvi I believe both of the problems demonstrated in your reproduction are related to the fact that `useLazyQuery` uses `client.watchQuery` behind the scenes, when it should just be using `client.query`:
* Refetching a `network-only` query when a variable changes is expected if the query is being watched, but would not happen if the query only executes when explicitly requested.
* The problem with `onCompleted` not firing is more loosely connected, but I'm sure that's something we could fix in the process of rewriting `useLazyQuery` to avoid watching, since we should be able to hook into the `Promise` returned by `client.query`.

Although we're about to release [v3.2.0](https://github.com/apollographql/apollo-client/pull/6774) with the changes that are currently on the `release-3.2` branch, we haven't forgotten about this issue. We hope to include a fix in a 3.2.x (or a 3.3 beta) version soon.

**Comment by mschipperheyn** [comment]
@benjamn I have a problem with useLazyQuery that I feel is similar and might be related to the cache. I execute a `useLazyQuery` based on an id and fragment that I know are already in the cache. The query never shoots off a network query (presumably bc the element is in the cache), loading never becomes true, neither the error key, but neither does it update with `data` key. The data key remains undefined. `onCompleted` is also never called. Setting the `fetchPolicy` to `network-only` or `no-cache` or setting `notifyOnNetworkStatusChange: true` has no effect. (tried from 3.1.5 to 3.2.4)

**Comment by vsylvestre** [comment]
I also have an issue with `useLazyQuery` that I think might relate to this?

I have an `ApolloClient` instance that I use in my `ApolloProvider` that I need to reinitialize once in a while because a couple of settings change during the lifetime of the app (mostly HTTP headers and connection params). Therefore, instead of having the client be created outside the scope of my React tree, it's generated within a custom hook. However, it seems like every time there's a new value passed to `ApolloProvider` for the `client` prop, all queries using `useLazyQuery` are relaunched automatically. I would expect that they wouldn't be launched until there's another call to the functions provided by the `useLazyQuery` hooks, right?

Maybe there's something wrong with the way I'm reloading the client though, and that it shouldn't go through React updates. I couldn't find any documentation on it. Let me know!

**Comment by hwharrison** [comment]
The bug calling onCompleted looks like a long-running holdover from 2.x.
https://github.com/apollographql/apollo-client/issues/6636

**Comment by 3nvi** [comment]
> Although we're about to release v3.2.0 with the changes that are currently on the release-3.2 branch, we haven't forgotten about this issue. We hope to include a fix in a 3.2.x (or a 3.3 beta) version soon.

@benjamn Given the fact that 3.3.0 launched without it, is there a chance we can get an update on the state of things? Is it something that's within your priorities or should we try and work around this for the time being.

As always, I appreciate you' re juggling a lot of things, so thanks in advance

**Comment by adrienharnay** [comment]
> As always, I appreciate you' re juggling a lot of things, so thanks in advance

:+1: on this, thanks for the work you're putting into this project!

**Comment by l222p** [comment]
Any update on this? I'm getting this error on version  

`"@apollo/client": "^3.2.5",`
`"apollo": "^2.31.1",`

**Comment by nadaru-fr** [comment]
Would be nice to have a proper onComplete fired on refetch even if it's from cache.

Meanwhile in order to handle it I manually check the cache before the call...

```
const queryDoc = gql`
  ...
`

function MyComponent() {
  const [query, queryResult] = useLazyQuery(queryDoc, {
    onCompleted: onCompleteFn,
  })
  const client = useApolloClient()

  const fnForRefetch = () => {
    const cached = client.readQuery({ query: queryDoc, variables })
      if (cached !== null) {
        onCompleteFn()
      } else {
        query({ variables })
      }
    }
  }
}
```

**Comment by kftsehk** [comment]
adding that `{loading}` doesn't include the code to process `onCompleted`, and will not fire `onCompleted` when another invocation (possibly with different variables) is in progress, making it hard to be compatible with any infinite scroll library.

```
"react-apollo": "^3.1.5",
```

**Comment by everdrone** [comment]
I'm having the same problems with `useLazyQuery`

**Comment by brainkim** [comment]
> @3nvi I believe both of the problems demonstrated in [your reproduction](https://github.com/apollographql/apollo-client/issues/6636#issuecomment-666310747) are related to the fact that `useLazyQuery` uses `client.watchQuery` behind the scenes, when it should just be using `client.query`:
> 
> * Refetching a `network-only` query when a variable changes is expected if the query is being watched, but would not happen if the query only executes when explicitly requested.
> * The problem with `onCompleted` not firing is more loosely related, but I'm sure that's something we could fix in the process of rewriting `useLazyQuery` to avoid watching, since we should be able to hook into the `Promise` returned by `client.query`.
> 
> Although we're about to release [v3.2.0](https://github.com/apollographql/apollo-client/pull/6774) with the changes that are currently on the `release-3.2` branch, we haven't forgotten about this issue. We hope to include a fix in a 3.2.x (or a 3.3 beta) version soon.

#

**Comment by brainkim** [comment]
THis should be fixed in 3.5, please try it.

**Comment by 3nvi** [comment]
@brainkim it's not. Unfortunately, I believe it's now worse than before, since `onCompleted` gets only triggered once for all lazy queries with `network-only` policy, regardless of the number of times they are called. This means that even if I perform multiple network requests, I'll still have only 1 `onCompleted` callback fired.

Please refer to this comment https://github.com/apollographql/apollo-client/issues/6636#issuecomment-666310747 and its reproduction instructions to validate the bugs. I was able to validate them in apollo-client@3.5.3.

Should we re-open this?

**Comment by brainkim** [comment]
> Unfortunately, I believe it's now worse than before

No, I would disagree with that. As far as I can tell, the original issue was about React render-looping when there was some combination of fetch policy and `onCompleted()` passed to `useQuery()`/`useLazyQuery()`. I’ve run through the examples listed in this issue and have yet to render-looping happen in 3.5. The other “bugs” listed throughout this issue are various expectations about when the `onCompleted()` callback should be called, but honestly I think this is underspecified by the documentation, which states:

> A callback function that's called when your query successfully completes with zero errors (or if errorPolicy is ignore and partial data is returned).
- https://www.apollographql.com/docs/react/api/react/hooks/#oncompleted

I’ll sift through this monster thread again to see if I can codify the general expectations as unit tests. Thanks for the patience.

**Comment by sbsupper** [comment]
This issue was not fixed for three years...

**Comment by mcsimps2** [comment]
@brainkim I'm new to using Apollo Client.  My "blank slate" expectation of reading the `onCompleted` section of the `useLazyQuery` documentation was that it would be called every time the query was run rather than just the first time.  If that wasn't the intention, then updating the docs would definitely save newcomers like myself a lot of digging through GitHub issues.

**Comment by apedroferreira** [comment]
i keep having to redo the authentication part of my site every time i update Apollo client and figuring out what changed with the `onCompleted` query handler... this is very confusing. 
i think the documentation needs improving and the behavior should be more consistent between non-major versions.

**Comment by meinto** [comment]
Does somebody know a drop in replacement for apollo/client?

**Comment by hwillson** [comment]
We believe this is now fixed; if anyone is still encountering this issue in `@apollo/client@latest`, definitely let us know. Thanks!
